File: src/main/java/bgu/spl171/net/impl/TFTPimplDown/BidiMessage.java
package bgu.spl171.net.impl.TFTPimplDown;


/**
 *  opcode operation
 *      1 Read request (RRQ)
 *      2 Write request (WRQ)
 *      3 Data (DATA)
 *      4 Acknowledgment (ACK)
 *      5 Error (ERROR)
 *      6 Directory listing request (DIRQ)
 *      7 Login request (LOGRQ)
 *      8 Delete request (DELRQ)
 *      9 Broadcast file added/deleted (BCAST)
 *      10 Disconnect (DISC)
 */
public class BidiMessage {

    private short opcode;
    private String fileName;
    private String userName;
    private short packetSize;
    private short blockNumber;
    private byte[] data;
    private byte deletedAdded;
    private short errorCode;
    private String errMsg;
    private byte aByte;

    public static BidiMessage createBcastMessage(int event, String fileName) {

        return new BidiMessage((short) 9, (byte) event, fileName, (byte) 0);
    }

    public static BidiMessage createAckMessage(int ackNum) {

        return new BidiMessage((short) 4, (short) ackNum);
    }

    public static BidiMessage createErrorMessage(int errNum, String errMsg) {

        return new BidiMessage((short) 7, (short) errNum, errMsg, (byte) 0);
    }

//    Constructors
    private BidiMessage(short opcode, String string, byte aByte) {
        this.opcode = opcode;
        switch (opcode){

            case 7: {
                this.userName = string;
            }

            case 1:
            case 2:
            case 8: {
                this.fileName = string;
            }


        }
        this.fileName = fileName;
        this.aByte = aByte;
    }

    private BidiMessage(short opcode) {

        this.opcode = opcode;
    }

    private BidiMessage(short opcode, short packetSize, short blockNumber, byte[] data) {

        this.opcode = opcode;
        this.packetSize = packetSize;
        this.blockNumber = blockNumber;
        this.data = data;
    }

    private BidiMessage(short opcode, short blockNumber) {

        this.opcode = opcode;
        this.blockNumber = blockNumber;
    }

    private BidiMessage(short opcode, byte deletedAdded, String fileName, byte aByte) {

        this.opcode = opcode;
        this.deletedAdded = deletedAdded;
        this.fileName = fileName;
        this.aByte = aByte;
    }

    private BidiMessage(short opcode, short errorCode, String errMsg, byte aByte) {

        this.opcode = opcode;
        this.errorCode = errorCode;
        this.errMsg = errMsg;
        this.aByte = aByte;
    }


//    getters
    public short getOpcode() {

        return opcode;
    }

    public String getFileName() {

        return this.fileName;
    }

    public byte getaByte() {

        return aByte;
    }

    public String getUserName() {
        return userName;
    }

    public short getPacketSize() {
        return packetSize;
    }

    public short getBlockNumber() {
        return blockNumber;
    }

    public byte[] getData() {
        return data;
    }

    public byte getDeletedAdded() {
        return deletedAdded;
    }

    public short getErrorCode() {
        return errorCode;
    }

    public String getErrMsg() {
        return errMsg;
    }
}

File: Client/src/BidiMessage.cpp

#include "../include/BidiMessage.h"


BidiMessage BidiMessage::createDataMessage(int packetSize, int blockNum, char* fileData) {

    return BidiMessage((short) 3, (short) packetSize, (short) blockNum, fileData);
}

BidiMessage BidiMessage::createBcastMessage(int event, string fileName){

    return BidiMessage((short) 9, (char) event, fileName, (char) 0);
}

BidiMessage BidiMessage::createAckMessage(int ackNum){

    return BidiMessage((short) 4, (short) ackNum);
}

BidiMessage BidiMessage::createErrorMessage(int errNum, string errMsg){

    return BidiMessage((short) 7, (short) errNum, errMsg, (char) 0);
}

BidiMessage BidiMessage::createLoginMessage(string userName) {

    return BidiMessage((short) 7, userName, (char) 0);
}


BidiMessage BidiMessage::createDeleteMessage(string fileName) {

    return BidiMessage((short) 8, fileName, (char) 0);
}

BidiMessage BidiMessage::createRRQMessage(string fileName) {

    return BidiMessage((short) 1, fileName, (char) 0);
}

BidiMessage BidiMessage::createWRQMessage(string fileName) {

    return BidiMessage((short) 1, fileName, (char) 0);
}

BidiMessage BidiMessage::createDirMessage() {

    return BidiMessage((short) 6);
}

BidiMessage BidiMessage::createDiscMessage() {

    return BidiMessage((short) 10);
}







//empty
BidiMessage::BidiMessage():
opcode(-1),
fileName(""),
userName(""),
packetSize(-1),
blockNumber(-1),
data((char*) ""),
deletedAdded((char) 0),
errorCode(-1),
errMsg(""),
aByte((char) 0),
complete(false),
bytesLength(0)
{}

//copy
BidiMessage::BidiMessage(const BidiMessage& original){

    this->opcode = original.getOpcode();
    std::cout << "isComplete original? "<< original.isComplete() << std::endl;
    this->complete = original.isComplete();
    std::cout << "isComplete? this"<< complete << std::endl;

    this->bytesLength = original.getBytesLength();
    switch (opcode){
//            LOGRQ
        case 7: {
            this->userName = original.getUserName();
            this->aByte = original.getaByte();
            break;
        }

//            DELRQ,RRQ,WRQ
        case 1:
        case 2:
        case 8: {
            this->fileName = original.getFileName();
            this->aByte = original.getaByte();
            break;
        }

//            DATA
        case 3: {

            this->packetSize = original.getPacketSize();
            this->blockNumber = original.getBlockNumber();
            this->data = original.getData();
            break;
        }

//            ACK
        case 4: {
            this->blockNumber = original.getBlockNumber();
            break;
        }

//            BCAST
        case 9: {

            this->deletedAdded = original.getDeletedAdded();
            this->fileName = original.getFileName();
            this->aByte = original.getaByte();
            break;
        }

//            ERROR
        case 5: {

            this->errorCode = original.getErrorCode();
            this->errMsg = original.getErrMsg();
            this->aByte  = original.getaByte();
            break;
        }

        default:
            break;
    }
}


BidiMessage::BidiMessage(short opcode, string aString, char aByte){

    this->opcode = opcode;
    this->complete = true;
    this->bytesLength = (int) aString.length() + 3;
    switch (opcode){

        case 7: {
            this->userName = aString;
            break;
        }

        case 1:
        case 2:
        case 8: {
            this->fileName = aString;
            break;
        }

        default:
            break;
    }
    this->fileName = aString;
    this->aByte = aByte;
}

BidiMessage::BidiMessage(short opcode):
opcode(opcode),
fileName(""),
userName(""),
packetSize(-1),
blockNumber(-1),
data((char*) ""),
deletedAdded((char) 0),
errorCode(-1),
errMsg(""),
aByte((char) 0),
complete(true),
bytesLength(2)
{}

BidiMessage::BidiMessage(short opcode, short packetSize, short blockNumber, char* data):
opcode(opcode),
fileName(""),
userName(""),
packetSize(packetSize),
blockNumber(blockNumber),
data(data),
deletedAdded((char) 0),
errorCode(-1),
errMsg(""),
aByte((char) 0),
complete(true),
bytesLength(packetSize+6)
{}

BidiMessage::BidiMessage(short opcode, short blockNumber):
opcode(opcode),
fileName(""),
userName(""),
packetSize(-1),
blockNumber(blockNumber),
data((char*) ""),
deletedAdded((char) 0),
errorCode(-1),
errMsg(""),
aByte((char) 0),
complete(true),
bytesLength(4)
{}

BidiMessage::BidiMessage(short opcode, char deletedAdded, string fileName, char aByte):
opcode(opcode),
fileName(fileName),
userName(""),
packetSize(-1),
blockNumber(-1),
data((char*) ""),
deletedAdded(deletedAdded),
errorCode(-1),
errMsg(""),
aByte(aByte),
complete(true),
bytesLength((int) fileName.length() + 4)
{}

BidiMessage::BidiMessage(short opcode, short errorCode, string errMsg, char aByte):
opcode(opcode),
fileName(""),
userName(""),
packetSize(-1),
blockNumber(-1),
data((char*) ""),
deletedAdded((char) 0),
errorCode(errorCode),
errMsg(errMsg),
aByte(aByte),
complete(true),
bytesLength((int) errMsg.length() + 5)
{}


//getters

short BidiMessage::getOpcode()const {

    return opcode;
}

string BidiMessage::getFileName()const {

    return fileName + "";
}

char BidiMessage::getaByte()const {

    return aByte;
}

string BidiMessage::getUserName()const {

    return userName + "";
}

short BidiMessage::getPacketSize()const {

    return packetSize;
}

short BidiMessage::getBlockNumber()const {

    return blockNumber;
}

char * BidiMessage::getData()const {

    char* dataCopy = (char*) "";

    for(int i=0; i<packetSize; i++){

        dataCopy[i] = this->data[i];
    }

    return dataCopy;
}

char BidiMessage::getDeletedAdded()const {

    return deletedAdded;
}

short BidiMessage::getErrorCode()const {

    return errorCode;
}

string BidiMessage::getErrMsg()const {

    return errMsg + "";
}

bool BidiMessage::isComplete() const {
    return complete;
}

int BidiMessage::getBytesLength() const {
    return bytesLength;
}


//setters

void BidiMessage::setOpcode(short opcode){

    this->opcode = opcode;
}

void BidiMessage::setFileName(string fileName){

    this->fileName = fileName;
}

void BidiMessage::setUserName(string userName){

    this->userName = userName;
}

void BidiMessage::setPacketSize(short packetSize){

    this->packetSize = packetSize;
}

void BidiMessage::setBlockNumber(short blockNumber){

    this->blockNumber = blockNumber;
}

void BidiMessage::setData(char* data){

    this->data = data;
}

void BidiMessage::setDeletedAdded(char deletedAdded){

    this->deletedAdded = deletedAdded;
}

void BidiMessage::setErrorCode(short errorCode){

    this->errorCode = errorCode;
}

void BidiMessage::setErrMsg(string errMsg){

    this->errMsg = errMsg;
}

void BidiMessage::setaByte(char aByte){

    this->aByte = aByte;
}

void BidiMessage::setComplete(bool complete) {
    BidiMessage::complete = complete;
}

void BidiMessage::setBytesLength(int byteLength) {
    BidiMessage::bytesLength = byteLength;
}

BidiMessage::~BidiMessage() {}

BidiMessage& BidiMessage::operator=(const BidiMessage& other) // copy assignment
{
    if (this != &other) { // self-assignment check expected

        this->opcode = other.getOpcode();
        this->fileName = other.getFileName();
        this->userName = other.getUserName();
        this->packetSize = other.getPacketSize();
        this->blockNumber = other.getBlockNumber();
        this->data = other.getData();
        this->deletedAdded = other.getDeletedAdded();
        this->errorCode = other.getErrorCode();
        this->errMsg = other.getErrMsg();
        this->aByte = other.getaByte();
        this->complete = other.isComplete();
        this->bytesLength = other.getBytesLength();
    }
    return *this;
}




